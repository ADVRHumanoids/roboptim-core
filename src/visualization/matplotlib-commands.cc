// Copyright (C) 2014 by Benjamin Chr√©tien, CNRS-LIRMM.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include <stdexcept>
#include "roboptim/core/config.hh"

#include <boost/format.hpp>
#include <boost/xpressive/xpressive.hpp>
#include <boost/preprocessor/stringize.hpp>

#include <roboptim/core/visualization/matplotlib-commands.hh>

namespace roboptim
{
  namespace visualization
  {
    namespace matplotlib
    {
      Import::Import (const std::string& packages)
        : command_ ("import " + packages)
      {}

      Import::Import (const std::string& from, const std::string& packages)
        : command_ ("from " + from + " import " + packages)
      {}

      Import::~Import ()
      {}

      Command::Command (const std::string& cmd)
	: command_ (cmd)
      {}

      Command::~Command ()
      {}

      const std::string&
      Command::command () const
      {
	return command_;
      }


# define MATPLOTLIB_UNARY_COMMAND(NAME)				\
      Command							\
      NAME ()							\
      {								\
	return Command ("plt." BOOST_PP_STRINGIZE(NAME) " ()");	\
      }

      Command
      comment (const char* content)
      {
	using namespace boost::xpressive;

	// Comment character
	std::string comment_char = "# ";

	// Newlines should be prefixed by #
	sregex pattern = sregex::compile ("(\n)");
	std::string format ("$1" + comment_char);

	return Command (comment_char +
			regex_replace (std::string (content), pattern, format));
      }

      Command
      set (const char* var, const char* value)
      {
        if (!*value)
	  {
	    throw std::runtime_error ("matplotlib::set (var, value) expects an "
				      "actual value.");
	  }

	boost::format fmt = boost::format ("%1% = %2%");

	// Ignore some irrelevant exceptions
	fmt.exceptions (boost::io::all_error_bits
			^ (boost::io::too_many_args_bit
			   | boost::io::too_few_args_bit
			   | boost::io::bad_format_string_bit));

	return Command ((fmt % var % value).str ());
      }

      MATPLOTLIB_UNARY_COMMAND (show)
      MATPLOTLIB_UNARY_COMMAND (figure)

# undef MATPLOTLIB_UNARY_COMMAND

    } // end of namespace matplotlib.
  } // end of namespace visualization.
} // end of namespace roboptim
