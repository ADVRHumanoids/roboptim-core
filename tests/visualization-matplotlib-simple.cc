// Copyright (C) 2014 by Benjamin Chr√©tien, CNRS-LIRMM.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "shared-tests/fixture.hh"

#include <iostream>

#include <roboptim/core/io.hh>

#include <roboptim/core/visualization/matplotlib.hh>
#include <roboptim/core/visualization/matplotlib-commands.hh>

using namespace roboptim;
using namespace roboptim::visualization;

BOOST_FIXTURE_TEST_SUITE (core, TestSuiteConfiguration)

BOOST_AUTO_TEST_CASE (visualization_matplotlib_simple)
{
  boost::shared_ptr<boost::test_tools::output_test_stream>
    output = retrievePattern ("visualization-matplotlib-simple");

  using namespace roboptim::visualization::matplotlib;
  Matplotlib matplotlib = Matplotlib::make_matplotlib (std::make_pair(1, 1), false);

  // Display nothing, generate just some valid matplotlib commands.
  (*output)
    << (matplotlib
	<< set ("foo", "42")
	<< comment ("Hello, world!\nThis is a three-line comment!\nAmazing!")
	<< import ("foo")
	<< import ("foo", "bar")
	<< figure ()
	<< show ()
	);

  // Test clear function.
  matplotlib.clear ();
  // Enable header.
  matplotlib.withHeader () = true;
  BOOST_CHECK (matplotlib.withHeader () == true);
  (*output) << std::endl << matplotlib;

  // Test some throws.
  BOOST_CHECK_THROW ((*output) << (matplotlib << set ("bar", 0)),
                     std::runtime_error);

  std::cout << output->str () << std::endl;
  BOOST_CHECK (output->match_pattern ());
}

BOOST_AUTO_TEST_SUITE_END ()
