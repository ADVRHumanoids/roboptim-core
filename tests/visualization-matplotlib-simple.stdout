foo = 42
# Hello, world!
# This is a three-line comment!
# Amazing!
fig = plt.figure()
plt.show()


#!/usr/bin/env python
# Generated by roboptim-core
import matplotlib
import matplotlib.pyplot as plt
import numpy as np

fig = plt.figure()
ax = plt.subplot(111)


box = ax.get_position()
ax.set_position([box.x0, box.y0, box.width * 0.8, box.height])
ax.legend(loc='center left', bbox_to_anchor=(1, 0.5))
fig.tight_layout()
plt.show()
